enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,284 @@
+The idea is to automate the manual steps that are required for consuming local storage in an OpenShift cluster by extending the Local Storage Operator (LSO).
+
+## Motivation
+
+To facilitate the consumption of locally-attached storage devices on OpenShift nodes in a cloud-native way.
+
+### Goals
+
+- Automatic detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Should respond to the attach/detach events of the disks/devices.
+- Should have options for filtering particular kind of disks based on properties such as name, size, manufacturer, etc.
+
+### Non-Goals
+
+## Proposal
+
+The `local-storage-operator` is already capable of consuming local disks from the nodes and provisioning PVs out of them,
+but the disk/device paths needs to explicitly specified in the `LocalVolume` CR.
+The idea is to introduce a new Custom Resource, `AutoDetectVolumes`, that enables automated discovery andn provisioning of local volume based PVs.
+
+### Risks and Mitigations
+
+- LSO will detect disks that contain data and/or are in-use via ensuring that the device:
+  - can be openened exclusively.
+  - is not read-only.
+  - is not removable.
+  - has no child partitions.
+  - has no FS signature.
+  - state (as outputted by `lsblk`) is not `suspended`
+- Ensuring disks aren't re-detected as new or otherwise destroyed if their device path changes.
+  - This is already ensured by the current LSO approach of consuming disks by their `UUID`
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```go
+
+type AutoDetectVolumeList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata"`
+    Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata"`
+    Spec              AutoDetectVolumeSpec   `json:"spec"`
+    Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+    // Nodes on which the automatic detection policies must run.
+    // +optional
+    NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+    // DeviceDiscoverPolicy is the policy that AutoDetectVolume's controller
+    // will be looking for during detection
+    DeviceDiscoveryPolicy DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicy,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+    Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+    // StorageClassName to use for set of matched devices
+    StorageClassName string `json:"storageClassName"`
+    // MinDeviceCount is the minumum number of devices that needs to be detected per node.
+    MinDeviceCount int `json:"minDeviceCount"`
+    // +optional
+    // Maximum number of Devices that needs to be detected per node.
+    MaxDeviceCount int `json:"maxDeviceCount"`
```

Comment:
> It's not clear how MinDeviceCount/MaxDeviceCount interacts between different `AutoDetectVolume` instances. Say I have three instances:
> 
> * A: requesting 2 block devices of any size
> * B: requesting 1 device of any size
> * C: requesting any number of devices of any size
> 
> And I have a node with 2 disks (sda, sdb). How are they going to be distributed among the `AutoDetectVolume` instances? Who decides the priority of each `AutoDetectVolume`?
> 

I believe each rule (AutoDetectVolume instance) will be computed in a FIFO manner. The operator watches for CRs and then reconciles them in the order they have been found I think.

> How to express "use 1 devices for OCS and all the other devices as regular LocalVolumes for applications"?
This could be achieved with 2 CRs, one for OCS (with its own rule) and one for the rest (own rule too).
Not sure what's bothering you here.

Reply:
FIFO basically mean that CRs are processed in random order. With informers and workqueues in multiple external components (LSO, local provisioner,local diskmaker) there is no guarantee about the order.

Reply:
I guess it doesn't rules matter if rules are properly written. If we don't see any devices for OCS when our CR is computed then this means other rules overrode it (most likely an admin error).

Reply:
This came up in designing `LocalVolume` CR too fwiw. Having multiple instances of `AutoDetectVolume` and `LocalVolume` that targets same node is going to be problematic.  Does it even make sense to allow user to target same node via different `AutoDetectVolume` instances?

Reply:
If someone targets the same node with multiple policies, it's their responsibility to ensure they are mutually exclusive, or else there will be undefined behavior. Again, with the flexibility comes the ability to shoot yourself in the foot.

Reply:
Agreed duplicate PVs will be coming up. I guess we can exclude the duplication by not considering consumed device paths for the PVs.

Reply:
IMO - you should not mark discussions like this as resolved until they have been really conclusively resolved. The reason being - marking a comment as resolved hides it from UI but anyone reading this enhancement in future should be aware that, this issue was discussed but not entirely resolved.